FsFilter1/FsFilter1.cpp

```diff
@@ -94,31 +108,138 @@ FLT_PREOP_CALLBACK_STATUS MiniPreWrite(PFLT_CALLBACK_DATA Data, PCFLT_RELATED_OB
 }
 
 
+NTSTATUS MiniConnect(PFLT_PORT clientport, PVOID serverportcookie, PVOID Context, ULONG size, PVOID *Connectioncookie) {
+	ClientPort = clientport;
+	TraceEvent(TRACE_VERBOSE, "Connect\r\n");
+	return STATUS_SUCCESS;
+}
+
+
+VOID MiniDisconnect(PVOID Connectioncookie) {
+	TraceEvent(TRACE_VERBOSE, "Disconnect\r\n");
+	FltCloseClientPort(FilterHandle, &ClientPort);
+}
+
+
+NTSTATUS MiniSendRecv(PVOID portcookie, PVOID InputBuffer, ULONG InputBufferLength, PVOID OutputBuffer, ULONG OutputBufferLength, PULONG RetLength) {
+	NTSTATUS status;
+	PCHAR msg = "KERNEL MSG";
+	TraceEvent(TRACE_INFO, "User msg: %s\r\n", (PCHAR)InputBuffer);
+	KdPrint(("User msg: %s\r\n", (PCHAR)InputBuffer));
+	strcpy((PCHAR)OutputBuffer, msg);
+	UNICODE_STRING TargetString;
+
+	TargetString.Buffer = (PWCH)ExAllocatePool(NonPagedPool, STRING_LEN * sizeof(WCHAR));
+	RtlZeroMemory(TargetString.Buffer, STRING_LEN * sizeof(WCHAR));
+	TargetString.MaximumLength = STRING_LEN;
+
+	if (strcmp((PCHAR)InputBuffer, "All drives") == 0) {
+		KdPrint(("ALL DRIVES\r\n"));
+	}
+	else {
+		KdPrint(("Individual Drives\r\n"));
+		wchar_t buf[STRING_LEN];
+		char inp_buf[STRING_LEN];
+		strcpy(inp_buf, "\\GLOBAL??\\");
+		strcat(inp_buf, (PCHAR)InputBuffer);
+		mbstowcs(buf, inp_buf, sizeof(inp_buf));
+		UNICODE_STRING symlinkName;
+		RtlInitUnicodeString(&symlinkName, buf);
+
+		OBJECT_ATTRIBUTES objectAttributes =
+			RTL_CONSTANT_OBJECT_ATTRIBUTES(&symlinkName,
+				OBJ_KERNEL_HANDLE | OBJ_CASE_INSENSITIVE);
+
+		HANDLE symLinkHandle = NULL;
+		status = ZwOpenSymbolicLinkObject(&symLinkHandle,
+			SYMBOLIC_LINK_ALL_ACCESS,
+			&objectAttributes);
+
+		if (!NT_SUCCESS(status)) {
+			if (symLinkHandle)
+				ZwClose(symLinkHandle);
+			ExFreePool(TargetString.Buffer);
+			KdPrint(("ZwOpenSymbolicLinkObject failed\n"));
+			return status;
+		}
+
+		ULONG Ret = 0;
+
+		status = ZwQuerySymbolicLinkObject(symLinkHandle,
+			&TargetString, &Ret);
+		if (!NT_SUCCESS(status)) {
+			if (symLinkHandle)
+				ZwClose(symLinkHandle);
+			ExFreePool(TargetString.Buffer);
+			KdPrint(("ZwQuerySymbolicLinkObject failed\n"));
+			return status;
+		}
+		ZwClose(symLinkHandle);
+	}
+
+	KdPrint(("AFTER Query SymLink %S\r\n", TargetString.Buffer));
+	if (!wcsncmp(TargetString.Buffer, L"\\Device", 7)) {
+		status = STATUS_SUCCESS;
+		KdPrint(("Success: TargetStr has Device\n"));
+	}
+	RtlZeroMemory(drive_filter.Buffer, STRING_LEN * sizeof(WCHAR));
+	RtlUnicodeStringCopy(&drive_filter, &TargetString);
+	ExFreePool(TargetString.Buffer);
+	return STATUS_SUCCESS;
+}
+
 NTSTATUS MiniUnload(FLT_FILTER_UNLOAD_FLAGS Flags)
 {
 	UNREFERENCED_PARAMETER(Flags);
 	
-	KdPrint(("Driver Unload.\r\n"));
+	TraceEvent(TRACE_INFO, "Driver Unload.\r\n");
+	FltCloseCommunicationPort(port);
```

Comment:
Can we call this on already closed port?

Reply:
Yes, may be we need NULL check for port before we Close the port.
And if it is closed, assign NULL to it.